File: miit_rule_checker/src/main/java/com/loper7/miit_rule_checker/LogHelper.java
package com.loper7.miit_rule_checker;

import android.nfc.Tag;
import android.text.TextUtils;
import android.util.Log;

import java.lang.reflect.Constructor;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

import top.canyie.pine.Pine;

/**
 * @Author loper7
 * @Date 2023/12/6 9:54
 * @Description 日志打印帮助
 **/
public class LogHelper {

    private static final String TAG = "MIIT_RULE_CHECKER";

    public static void e(String message) {
        Log.e(TAG, message);
    }

    public static void printHookedMethod(String name) {
        Log.w(TAG, "----------------------" + TAG + "----------------------");
        Log.w(TAG, "*****Method call detected*****");
        Log.w(TAG, "    " + name);
        Log.w(TAG, "******************************");
        Log.w(TAG, "Stack information:");
        Log.w(TAG, getMethodStack());
        Log.w(TAG, "----------------------" + TAG + "----------------------");
    }

    public static void printMethodCount(HashMap<String, HashMap<String, Integer>> map) {
        Log.w(TAG, "----------------------" + TAG + "----------------------");
        if (map == null || map.size() == 0) {
            Log.i(TAG, "设定时间内指定方法没有被调用");
        } else {
            for (Map.Entry<String, HashMap<String, Integer>> entry : map.entrySet()) {
                Log.d(TAG, ">>>>>>>>>>> " + entry.getKey());
                for (Map.Entry<String, Integer> entry1 : entry.getValue().entrySet()) {
                    Log.i(TAG, "stackClassName: " + entry1.getKey() + "\tcount:" + entry1.getValue());
                }
                Log.d(TAG, "");
            }
        }
        Log.w(TAG, "----------------------" + TAG + "----------------------");
    }


    public static String getMethodName(Pine.CallFrame callFrame) {
        if (callFrame == null) return "";
        Constructor<?>[] constructors = callFrame.method.getDeclaringClass().getDeclaredConstructors();
        String name = "";
        if (constructors.length > 0) {
            try {
                name = String.format("%s.%s", constructors[0].toString().split(" ")[1].split("\\(")[0], callFrame.method.getName());
            } catch (Exception e) {
                name = String.format("%s.%s", constructors[0], callFrame.method.getName());
            }
        }
        return name;
    }

    public static String getMethodStack() {
        StackTraceElement[] stackTraceElements = Thread.currentThread().getStackTrace();
        StringBuilder builder = new StringBuilder();
        for (StackTraceElement element : stackTraceElements) {
            String className = element.getClassName();
            String currPackage = Objects.requireNonNull(LogHelper.class.getPackage()).getName();
            boolean isChecker = TextUtils.isEmpty(className) ||
                    className.startsWith("top.canyie.pine") ||
                    className.startsWith("dalvik.system.VMStack") ||
                    className.startsWith("java.") ||
                    className.startsWith(currPackage);
            if (isChecker) continue;
            builder.append(element.toString());
            builder.append("\n");
        }
        return builder.toString();
    }

    public static String getStackClassName() {
        StackTraceElement[] stackTraceElements = Thread.currentThread().getStackTrace();
        for (StackTraceElement element : stackTraceElements) {
            String className = element.getClassName();
            String currPackage = Objects.requireNonNull(LogHelper.class.getPackage()).getName();
            boolean isChecker = TextUtils.isEmpty(className) ||
                    className.startsWith("top.canyie.pine") ||
                    className.startsWith("dalvik.system.VMStack") ||
                    className.startsWith("java.") ||
                    className.startsWith(currPackage);
            if (isChecker) continue;
            String[] splits = className.split("\\$");
            return splits[0];
        }
        return "";
    }

}

File: miit_rule_checker/src/main/java/com/loper7/miit_rule_checker/LogHelper.kt
package com.loper7.miit_rule_checker

import android.text.TextUtils
import android.util.Log
import top.canyie.pine.Pine
import java.lang.reflect.Member

/**
 *@Author loper7
 *@Date 2023/11/28 17:54
 *@Description
 **/
object LogHelper {
    private const val TAG = "MIIT_RULE_CHECKER"


    fun e(message : String?) = Log.e(TAG , message ?: "")


    fun printHookedMethod(name : String) {
        Log.w(TAG , "----------------------${TAG}----------------------")
        Log.w(TAG , "*****Method call detected*****")
        Log.w(TAG , "    ${name}")
        Log.w(TAG , "******************************")
        Log.w(TAG , "Stack information:")
        Log.w(TAG , getMethodStack())
        Log.w(TAG , "----------------------${TAG}----------------------")
    }

    fun printMethodCount(map : HashMap<String , HashMap<String , Int>>?) {

        Log.w(TAG , "----------------------${TAG}----------------------")
        if (map.isNullOrEmpty()) {
            Log.i(TAG , "设定时间内指定方法没有被调用")
        } else {
            map.forEach {
                Log.d(TAG , ">>>>>>>>>>> ${it.key}")
                it.value.forEach { value ->
                    Log.i(TAG , "stackClassName: ${value.key}\tcount: ${value.value}")
                }
                Log.d(TAG , "")
            }
        }
        Log.w(TAG , "----------------------${TAG}----------------------")
    }

    fun getMethodName(callFrame : Pine.CallFrame?) : String {
        if (callFrame == null) return ""
        val constructors = callFrame.method.declaringClass.declaredConstructors
        val name = if (constructors != null && constructors.isNotEmpty()) {
            try {
                "${constructors[0].toString().split(" ")[1].split("(")[0]}.${callFrame.method.name}"
            } catch (e : Exception) {
                "${constructors[0]}.${callFrame.method.name}"
            }
        } else {
            "${callFrame.method.declaringClass.`package`?.name}.${callFrame.method.name}"
        }
        return name
    }

    fun getMethodStack() : String {
        val stackTraceElements = Thread.currentThread().stackTrace
        val builder = StringBuilder()
        for (element in stackTraceElements) {
            val line = element.toString()
            val className = element.className
            val currPackage = this.javaClass.`package`?.name ?: "com.loper7.miit_rule_checker"
            val isChecker =
                TextUtils.isEmpty(className) || className.startsWith("top.canyie.pine") || className.startsWith("dalvik.system.VMStack") || className.startsWith(
                    "java.") || className.startsWith(currPackage)
            if (isChecker) continue
            builder.append(line)
            builder.append("\n")
        }
        return builder.toString()
    }

    fun getStackClassName() : String {
        val stackTraceElements = Thread.currentThread().stackTrace
        for (element in stackTraceElements) {
            val className = element.className
            val currPackage = this.javaClass.`package`?.name ?: "com.loper7.miit_rule_checker"
            val isChecker =
                TextUtils.isEmpty(className) || className.startsWith("top.canyie.pine") || className.startsWith("dalvik.system.VMStack") || className.startsWith(
                    "java.") || className.startsWith(currPackage)
            if (isChecker) continue
            val splits = className.split("\\$".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            return splits[0]
        }
        return ""
    }

}